module.exports = class HandlersProduct {
    constructor(productUseCases) {
        this.productUseCases = productUseCases;
    }

    getProductHandler = async (req, res) => {
        try {
            const { id } = req.params;
            const {
                message,
                code,
                data,
                error = null,
            } = await this.productUseCases.getOneProductUseCase(
                parseInt(id, 10),
            );
            if (code >= 400) return res.status(code).send({ message, error });
            return res.status(code).send({
                message,
                data,
            });
        } catch (error) {
            return res.status(500).send({
                code: 'fail',
                message: 'there was an internal error',
            });
        }
    };

    getProductsHandler = async (req, res) => {
        try {
            const { message, code } = await this.productUseCases.getProductsUseCase(
                req.query,
            );
            if (code >= 400) return res.status(code).send(message);
            return res.status(code).send({
                ...message,
            });
        } catch (error) {
            return res.status(500).send({
                code: 'fail',
                message: 'there was an internal error',
            });
        }
    };

    postProductHandler = async (req, res) => {
        try {
            const {
                message,
                code,
                error = null,
            } = await this.productUseCases.createProductUseCase(
                req.body,
            );
            if (code >= 400) return res.status(code).send({ message, error });
            return res.status(code).send({
                data: message,
            });
        } catch (error) {
            return res.status(500).send({
                code: 'fail',
                message: 'there was an internal error',
            });
        }
    };

    putProductHandler = async (req, res) => {
        try {
            const { id } = req.params;
            const {
                message,
                code,
                data,
                error = null,
            } = await this.productUseCases.updateProductUseCase(req.body, id);

            if (code === 400) return res.status(code).send({ message, error });
            return res.status(code).send({
                message,
                data,
            });
        } catch (error) {
            return res.status(500).send({
                code: 'fail',
                message: 'there was an internal error',
            });
        }
    };

    deleteProductHandler = async (req, res) => {
        try {
            const { id } = req.params;
            const {
                code,
                data,
            } = await this.productUseCases.deleteProductUseCase(id);

            return res.status(code).send({ data });
        } catch (error) {
            return res.status(500).send({
                code: 'fail',
                message: 'there was an internal error',
            });
        }
    };
};
